### PR TITLE
Introduce configuration for "Go to Symbol in Workspace"

### PR DESCRIPTION
**Pull request by dmgonch**
Allow configuring number of chars user must to type in for "Go to Symbol in Workspace" command to return any results (default is 0 to preserve existing behavior). The reasoning here is that when lots of projects are loaded by OmniSharp it makes little sense to return all symbols discovered. This change should largely address these issues as well: https://github.com/OmniSharp/omnisharp-roslyn/issues/1243, https://github.com/OmniSharp/omnisharp-vscode/issues/1808.
Additionally allow configuring max number of items returned by "Go to Symbol in Workspace" command. Similarly to the above in a big repo returning pages and pages of items in most cases will be wasteful. But default is 0 which is indicates no limit to preserve existing behavior.

### CODE REVIEW

**Comment by dmgonch** [comment]
✓ Prints the download status to the logger as Downloading package 'somePackage' (4 KB)..........Failed

Building locally without any of my changes produces the same failure for me. The build is flaky? Is there a way to kick off a retry?

**Comment by SirIntruder** [comment]
Looking forward to fixing this 👍 

Question: Should "minFindSymbolsFilterLength" wait for server respone at all? Extension could skip sending request and return empty array immediately. IIRC Visual Studio by default waits for one character, I assume this should be the default value here as well, instead of zero.

**Comment by dmgonch** [comment]
@SirIntruder : 
- setting minFindSymbolsFilterLength to 1 by default make sense but I was going for back compat unless repo owners decide otherwise.
- technically https://github.com/OmniSharp/omnisharp-roslyn/blob/6836fadb9c35a88d4695276d14302336a460b841/src/OmniSharp.Host/Endpoint/EndpointHandler.cs#L176 aggregates requests from all handlers so again I went for back compat though practically I think you are right and this call can be avoided altogether (though it is a quite cheap one). Will wait for more feedback one way or another.

**Comment by akshita31** [comment]
@dmgonch That is the first time I have seen that test being flaky. For now I merged master and the test passes.

File: src/observers/OptionChangeObserver.ts [hunk 1]
```diff
@@ -33,5 +33,7 @@ export function ShowOmniSharpConfigChangePrompt(optionObservable: Observable<Opt
 function hasChanged(oldOptions: Options, newOptions: Options): boolean {
     return (oldOptions.path != newOptions.path ||
         oldOptions.useGlobalMono != newOptions.useGlobalMono ||
+        oldOptions.minFindSymbolsFilterLength != newOptions.minFindSymbolsFilterLength ||
```
[the remark this answers is not in the document]
Comment (dmgonch) [reply src/observers/OptionChangeObserver.ts]: added in the last update

File: src/omnisharp/server.ts [hunk 2]
```diff
@@ -306,6 +306,14 @@ export class OmniSharpServer {
             args.push('--debug');
         }
 
+        // Only add arguments for values that are different from defaults and are meaningful
+        if (options.minFindSymbolsFilterLength > 0) {
+            args.push(`FindSymbols:MinFilterLength=${options.minFindSymbolsFilterLength.toString()}`);
```
[the remark this answers is not in the document]
Comment (dmgonch) [reply src/omnisharp/server.ts]: indeed the experience is my friendlier. please see the last update.

File: src/omnisharp/options.ts [hunk 3]
```diff
@@ -63,6 +65,9 @@ export class Options {
 
         const disableMSBuildDiagnosticWarning = omnisharpConfig.get<boolean>('disableMSBuildDiagnosticWarning', false);
 
+        const minFindSymbolsFilterLength = omnisharpConfig.get<number>('minFindSymbolsFilterLength', 0);
+        const maxFindSymbolsItems = omnisharpConfig.get<number>('maxFindSymbolsItems', 1000); // <=0 means 'Unlimited'
```
[the remark this answers is not in the document]
Comment (dmgonch) [reply src/omnisharp/options.ts]: Yes, it does. Why? What is the confusion?

File: src/observers/OptionChangeObserver.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmgonch) [reply src/observers/OptionChangeObserver.ts]: Sorry, could you please clarify? the test I thought checks that when options change the observer detects this. These settings are still in the settings.josn - it just they take effect without needing restarting VSCode.

[the remark this answers is not in the document]
Comment (akshita31) [reply src/observers/OptionChangeObserver.ts]: Sorry, I think the code here maybe confusing. This file here is just to detect that if `path`, `globalMono` or `waitForDebugger` has changed and if yes them create a prompt for the user to restart the omnisharp process. This is a special case and has nothing to do with the normal option stream.
As per your code, whenever a user creates a "Go to Symbol" request, you will read the "latest" options using the optionProvider and accordingly service the request, there is no need to prompt the user to restart(in short no need to touch this file here at all). Does that make sense ?

File: test/unitTests/OptionObserver/OptionChangeObserver.test.ts [hunk 4]
```diff
@@ -41,6 +41,8 @@ suite("OmniSharpConfigChangeObserver", () => {
     [
         { config: "omnisharp", section: "path", value: "somePath" },
         { config: "omnisharp", section: "waitForDebugger", value: true },
+        { config: "omnisharp", section: "minFindSymbolsFilterLength", value: 5 },
```
[the remark this answers is not in the document]
Comment (akshita31) [reply test/unitTests/OptionObserver/OptionChangeObserver.test.ts]: We don't need any change in the OptionChangeObserver. To detect whether we are able to handle the change in the options, you should add a test to the optionProvider.tests

File: src/observers/OptionChangeObserver.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmgonch) [reply src/observers/OptionChangeObserver.ts]: Got it - makes sense. I will update the observer not to detect the change. Thanks for clarifying!

File: src/omnisharp/options.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (akshita31) [reply src/omnisharp/options.ts]: No the placement of the code here is correct. I just want to understand about the comment here. I believe the comment was added when we were taking the default value of 'maxFindSymbolsItems' to be 0. But now that we are taking the default value to be 1000, why do you think the user will enter a value less than 0?

[the remark this answers is not in the document]
Comment (akshita31) [reply src/omnisharp/options.ts]: Also, I am not sure whether it is possible in the settings, but we could add some restriction like should be greater than 0 using the json schema validation. https://cswr.github.io/JsonSchema/spec/basic_types/#numeric-schemas

File: test/unitTests/OptionObserver/OptionChangeObserver.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmgonch) [reply test/unitTests/OptionObserver/OptionChangeObserver.test.ts]: reverted the change.

**Comment by dmgonch** [comment]
@akshita31 Since I didn't make any changes the "Option Provider" class itself, could you please clarify what tests you would like to see added to OptionProvider.test.ts? 

Thanks for signing off but these changes might need to be reworked based on the outcome of https://github.com/OmniSharp/omnisharp-roslyn/pull/1284#discussion_r217105725

**Comment by akshita31** [comment]
@dmgonch The optionProvider has a function call "GetLatestOptions" which returns the latest options when the options are changed, so you might add a test there that changes your options and checks if on invoking the latest options you are getting the required changes. But yes you are right you didn't make any change to that class so doesnt make sense to add the test there. Thanks for clarifying.

**Comment by dmgonch** [comment]
@rchande : do you still have any concerns about this change?

**Comment by dmgonch** [comment]
@rchande , @akshita31: Rebuild failed with another error seeming unrelated to my change "The command "gulp 'vsix:release:package'" failed". What should I do to get pass the failure? Still cannot find a how to restart a build manually.

**Comment by akshita31** [comment]
@dmgonch I have seen this problem when there is some error when I execute "npm run compile". Can you check if that is the case here ?

**Comment by dmgonch** [comment]
@akshita31 : fixed - thanks! Any suggestions on the coverage failure?

**Comment by dmgonch** [comment]
@akshita31 : I might be reading the report incorrectly but it looks like all lines that I added are covered, no? https://codecov.io/gh/OmniSharp/omnisharp-vscode/compare/43b6fb24458787e7bd65f0c5b7df129aa5f89832...1c8828975ab778dba8acee66014b6347c48b5eef/diff

**Comment by akshita31** [comment]
@dmgonch You are right, the coverage failure is probably because we do not have any integration/unit tests that cover the WorkspaceSymbolProvider. But the coverage is not a mandatory test, so I will go ahead and merge your PR now. If you are interested you can create a PR to add an integration test for the same,otherwise I will take it up as a separate work item myself. Also, you can add a change log update to both the repositories.

Thanks for your contribution

**Pull request merged**